### PR TITLE
Search products by SKU: add a segmented control to products search

### PR DESCRIPTION
**Pull request by jaclync**
Part of #3157 
Based on https://github.com/woocommerce/woocommerce-ios/pull/7738

### Description

To support searching products by SKU, we need to introduce a segmented control for the user to manually select whether they want to search by SKU or not due to performance reason (more details in p91TBi-96v-p2#comment-10714). To make it work with the existing shared search UI, I had to add 3 new things to the `SearchUICommand` protocol in the implementation that I came up with last HACK Week:

```swift
/// A closure to resynchronize models if the data source might change (e.g. when the filter changes in products search).
var resynchronizeModels: (() -> Void) { get set }

/// The optional view to show between the search bar and search results table view.
/// If `nil`, the search bar is right above the search results.
func createHeaderView() -> UIView?

/// The predicate to fetch product search results based on the keyword.
/// Called when a search API request is made for the keyword.
/// - Parameter keyword: search query.
/// - Returns: predicate that is based on the search keyword. When the keyword is empty, `nil` can be returned as an example use case.
func searchResultsPredicate(keyword: String) -> NSPredicate?
```

The high-level changes are:
- UI: A new view with a segmented control ("All Products" and "SKU") is added in the `ProductSearchUICommand.createHeaderView` implementation
- When the merchant taps on a different filter, `ProductSearchUICommand.resynchronizeModels` closure is called to resync the search results. When the SKU filter is selected, the remote uses the `search_sku` parameter that works for stores with WC version 6.6+. If the store WC version is not eligible for partial SKU search, it does a full SKU search using the `sku` parameter to find any products with the exact SKU
- Internally, a dictionary from a filter to the latest search query is kept so that we only make an API request once per query and filter for the first page
- The changes are behind a new feature flag `searchProductsBySKU`

Apology for the 500+ diffs, the network response mock JSON is 82 diffs and the test cases need to cover when the feature flag is disabled. Please share any other implementation ideas to simplify the changes!

#### WC store with version 6.6+

- Log in if needed, and connect to a store with WC 6.6+
- Go to the Products tab
- Tap on the search icon --> there should be a segmented control between the search bar and the results, and a list of existing products should be shown for both tabs
- Type in some query --> the results should be shown correctly for each tab, where products with a partial match on the SKU should be shown on the SKU tab. pagination should work correctly on both tabs
- With the same query, go to another tab and then back --> the results should stay the same (no extra loading)

#### WC store with version below 6.6

- Log in if needed, and connect to a store with WC version below 6.6
- Go to the Products tab
- Tap on the search icon --> there should be a segmented control between the search bar and the results, and a list of existing products should be shown for both tabs
- Type in some query --> the results should be shown correctly for the All Products tab. for the SKU tab, only products (usually up to one unless some plugins are installed that allow duplicate SKUs) with the same SKU as the query should be shown
- With the same query, go to another tab and then back --> the results should stay the same (no extra loading)

#### Feature flag off

- Disable the `searchProductsBySKU` feature flag
- Log in if needed
- Go to the Products tab
- Tap on the search icon --> there should not be a segmented control between the search bar and the results, the UX should be the same as in production except for the search bar placeholder
- Type in some query --> the results should be shown the same way as in production, pagination should work as before

### Screenshots

Empty query | SKU results
-- | --
Simulator Screen Shot - iPhone 13 Pro - 2022-09-16 at 16 32 47 | Simulator Screen Shot - iPhone 13 Pro - 2022-09-16 at 16 32 57

#### Feature flag off

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by itsmeichigo — COMMENTED** [review]
Great work - the SKU search works as expected for me 💯  

There's this existing issue though - it seems that when we select a filter from the product list, the search list doesn't work very consistently. It will first show the results with the filter, and after clearing the keyword, the filter seems to be ignored. I tested on the current version of the app and found a similar behavior:


https://user-images.githubusercontent.com/5533851/190973827-7c80e45b-dd25-423f-8dc9-87dcb9b844b3.mp4

Also - it seems that the additions to the `SearchUICommand` protocol cause the other classes that conform it to change too. We already have [`ProductSelector`](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelector.swift) for searching products (used in Order creation and Coupons). Do you think we should migrate the product search screen there to avoid making changes to `SearchUICommand`?

**Comment by jaclync** [comment]
> There's this existing issue though - it seems that when we select a filter from the product list, the search list doesn't work very consistently. It will first show the results with the filter, and after clearing the keyword, the filter seems to be ignored. I tested on the current version of the app and found a similar behavior:

I think it makes sense to show all the products when the search query is empty, in case the product that the merchant is looking for is hidden from the filters on the products tab. Do you think this is a blocker? Somehow I can't reproduce the temporary SKU results at 0:18 in your screencast with your test site, but I added a commit https://github.com/woocommerce/woocommerce-ios/pull/7739/commits/267e69a5f92944fddc355010c9de572a357e83b8 for the FRC to perform fetch after the predicate changes (the function is only triggered when the search query changes with debounce).

> Also - it seems that the additions to the SearchUICommand protocol cause the other classes that conform it to change too. We already have [ProductSelector](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelector.swift) for searching products (used in Order creation and Coupons). Do you think we should migrate the product search screen there to avoid making changes to `SearchUICommand`?

Thanks for mentionning this, I didn't know that we have another implementation of products search in SwiftUI! It'd be nice to migrate it to SwiftUI, but I don't think it's going to be a trivial task. Since this is a feature already in Android, I think it'd be great if we can ship it in iOS and then refactor it to SwiftUI in the future (maybe my next HACK Week in Q4). I'll create an issue for the migration if you agree.

I'll only merge the storage PR https://github.com/woocommerce/woocommerce-ios/pull/7738 when this PR is good to go.

File: WooCommerce/Classes/ViewRelated/Search/Product/ProductSearchUICommand.swift [hunk 1]
```diff
@@ -71,4 +126,36 @@ final class ProductSearchUICommand: SearchUICommand {
...
+
+    func searchResultsPredicate(keyword: String) -> NSPredicate? {
+        guard isSearchProductsBySKUEnabled else {
+            return NSPredicate(format: "ANY searchResults.keyword = %@", keyword)
+        }
+        guard keyword.isNotEmpty else {
+            return nil
+        }
+        return NSPredicate(format: "SUBQUERY(searchResults, $result, $result.keyword = %@ AND $result.filterKey = %@).@count > 0",
+                           keyword, filter.rawValue)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Search/Product/ProductSearchUICommand.swift:138]: #TIL 👏

File: WooCommerce/Classes/ViewRelated/Search/SearchViewController.swift [hunk 2]
```diff
@@ -368,6 +386,13 @@ private extension SearchViewController {
                 self?.synchronizeSearchResults(with: query)
             }
     }
+
+    func configureSearchResync() {
+        searchUICommand.resynchronizeModels = { [weak self] in
+            guard let self = self else { return }
+            self.synchronizeSearchResults(with: self.searchQuery)
+        }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Search/SearchViewController.swift:394]: ❓ Will this override the variable set in the `searchUICommand`? If so the command will lose its initial variable, should we keep it as readonly in the protocol?

**Review by itsmeichigo — COMMENTED** [review]
> I think it makes sense to show all the products when the search query is empty

What I meant in that comment is that the list shows only filtered products (with status draft) initially, and then after the search query is updated, we're showing all results disregarding the status. Should we ignore the filtered status initially too?

> Since this is a feature already in Android, I think it'd be great if we can ship it in iOS and then refactor it to SwiftUI in the future (maybe my next HACK Week in Q4). I'll create an issue for the migration if you agree.

Yes, that makes sense!

File: WooCommerce/Classes/ViewRelated/Search/SearchViewController.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Search/SearchViewController.swift:394]: this is the only place where the actual resync logic is implemented, I had to make it a callback variable in order for the command to call it when the results need to be resynced. I'm open to any other ideas!

also added a comment on the closure in the protocol in https://github.com/woocommerce/woocommerce-ios/pull/7739/commits/065b3b6ca9fe7f9338ba0fccc559436fa7ee034a

**Comment by jaclync** [comment]
> What I meant in that comment is that the list shows only filtered products (with status draft) initially, and then after the search query is updated, we're showing all results disregarding the status. Should we ignore the filtered status initially too?

Oh I see now. The filtered products are the initial results because they are the only products in storage synced from the products tab with the filters, and we don't make another API request when launching the products search screen. It is quite a dilemma whether we want to make the initial state consistent either by making an API request to fetch all products (not ideal due to performance reasons) or also applying the filters (which might not be the best UX either). Maybe we can see if any merchants mention any issues and update the feature?

**Review by itsmeichigo — APPROVED** [review]
> Maybe we can see if any merchants mention any issues and update the feature?

Agreed 🙂

**Pull request merged**